### PR TITLE
fix(background): update icon change logic; using TabState for `sessions`

### PR DESCRIPTION
**Pull request by sidvishnoi**
## Context

Closes https://github.com/interledger/web-monetization-extension/issues/381

This PR got a bit bigger, with related/unrelated changes. Can split if wanted.

- Moves `sessions` to `TabState` service, adding some utils
- Use tab state to sync icons and tooltip text; some refactoring around that
   - Also remove reliance on `SET_IS_SITE_MONETIZED` message event (moved into SendToPopup)
- SendToPopup: save last message that wasn't sent to popup as it was closed; send it as soon as popup opens.

### CODE REVIEW

**Comment by raducristianpopa** [comment]
<h2>Extension builds preview</h2>

<table role="table">
<thead>
  <tr>
    <th>Name</th>
    <th>Link</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td align="center">Latest commit</td>
    <td align="center">ec3d139dfdecaaea8ade2a50359874d3c995198a</td>
  </tr>
  <tr>
    <td align="center">Latest job logs</td>
    <td align="center"><a href="https://github.com/interledger/web-monetization-extension/actions/runs/9909467613">Run #9909467613</a></td>
  </tr>
  <tr><td align="center">Badge</td><td align="center"><a href="https://github.com/interledger/web-monetization-extension/suites/25941522312/artifacts/1695503110">Download</a></td></tr><tr><td align="center">Badge</td><td align="center"><a href="https://github.com/interledger/web-monetization-extension/suites/25941522312/artifacts/1695503112">Download</a></td></tr>
</tbody>
</table>

File: src/background/services/tabState.ts [hunk 1]
```diff
@@ -56,4 +58,11 @@ export class TabState {
       state.lastPaymentTimestamp = now
     }
   }
+
+  getSessions(tab: Tabs.Tab) {
+    if (!this.sessions.has(tab)) {
+      this.sessions.set(tab, new Map())
```
Comment (sidvishnoi) [inline src/background/services/tabState.ts:64]: Even though we're setting in a "get", it helps to keep code clean. Initializing empty Map shouldn't be expensive (in terms of memory or CPU).

Comment (raducristianpopa) [reply src/background/services/tabState.ts:64]: This is exactly about what I was thinking right now. It looks like whenever we will change the focus from one tab  to another we will perform the `set` (one for `resume` and one for `pause`). The reason I believe we should avoid this, is that people (me included), have a lot of tabs open and we will rely on the service worker to be suspended to actually free up the memory.

Comment (sidvishnoi) [reply src/background/services/tabState.ts:64]: As the key is `Tab`, memory freeing isn't an issue though.
Empty Map doesn't take much memory.

**Comment by sidvishnoi** [comment]
There's some issue in this approach. It keeps showing "this site is not monetized". Will investigate.

**Comment by sidvishnoi** [comment]
> There's some issue in this approach. It keeps showing "this site is not monetized". Will investigate.

@raducristianpopa 🤦🏽 
```ts
const t1 = await this.getTabById(tabId)
const t2 = await this.getTabById(tabId)
console.log(t1 === t2) // false
```

**Comment by sidvishnoi** [comment]
Blocked by https://github.com/interledger/web-monetization-extension/issues/286

File: src/background/services/tabEvents.ts [hunk 2]
```diff
@@ -69,34 +83,60 @@ export class TabEvents {
...
+    enabled,
+    state,
+    isTabMonetized
+  }: {
+    enabled: Storage['enabled']
+    state: Storage['state']
+    tabId?: TabId
+    isTabMonetized?: boolean
+  }) {
+    isTabMonetized ??= tabId ? this.tabState.getSessions(tabId).size > 0 : false
```
Comment (raducristianpopa) [inline src/background/services/tabEvents.ts:131]: Not to keen about mutating the parameter.

**Review by raducristianpopa — COMMENTED** [review]
Can't we remove the `IS_TAB_MONETIZED` message from content script to background at this point?

**Review by raducristianpopa — APPROVED** [review]

**Comment by sidvishnoi** [comment]
> Can't we remove the `IS_TAB_MONETIZED` message from content script to background at this point?

Not yet, as we need to handle on-page link changes too. Will do as follow-up.

**Pull request merged**